saq/queue.py

```diff
@@ -393,12 +412,122 @@ async def enqueue(self, job_or_func, **kwargs):
         job.queued = now()
         job.status = Status.QUEUED
 
-        if not await self._enqueue_script(
+        on_enqueue_generators = [
+            on_enqueue(job) for on_enqueue in self._on_enqueue.values()
+        ]
+        for gen in on_enqueue_generators:
+            await run_iterator_to_first_yield(gen)
```

Comment:
why is this done? first_yield and then later run iterator to completion?

Reply:
This let's users write callbacks that look like this:


```python
async def on_enqueue(job):
    state = await before()
    try:
        yield
    finally:
        await after(state)
```

That makes it really easy to share state between before and after callbacks. Otherwise you have to mess with ContextVars, monkeypatching jobs, or some other messy techniques.

I don't think we strictly need any `after` callbacks yet. So maybe we should just remove this in favor of a single `before_enqueue`?